<?xml version="1.0"?>
<!--
  This file is supposed to launch the drivers of the real robots instead of the simulation.
-->
<launch>

  <!-- Required for the UR scripts to be sent -->
  <param name="use_real_robot" type="bool" value="true" />

  <!-- robot_ip: IP-address of the robot's socket-messaging server -->
  <arg name="robot_ip_a"  default="10.66.171.52"/>
  <arg name="robot_ip_b"  default="10.66.171.51"/>
  <!-- <arg name="limited"	  default="false"/> -->
  <arg name="limited"	  default="true"/>
  <arg name="min_payload" default="0.0"/>
  <arg name="max_payload" default="1.0"/>
  <arg name="servoj_time" default="0.008" />
  <arg name="shutdown_on_disconnect" default="true" />
  <!-- require_activation defines when the service /ur_driver/robot_enable needs to be called. -->
  <arg name="require_activation" default="Never" /> <!-- Never, Always, OnStartup -->

  <!-- The max_velocity parameter is only used for debugging in the ur_driver. It's not related to actual velocity limits -->
  <arg name="max_velocity" default="10.0"/> <!-- [rad/s] -->

  <!-- suction controller parameters -->
  <arg name="config_dir"        default="$(find o2as_fastening_tools)/config"/>
  <arg name="suction_control"   default="suction_control"/>
  <arg name="operation_mode"    default="operation_mode"/>

  <!-- a_bot realsense parameters -->
  <arg name="trigger_mode" default="false"/>
  <arg name="send_color"   default="true"/>
  <arg name="send_depth"   default="true"/>
  <arg name="send_cloud"   default="true"/>


  <!-- The three lines below are commented out because it is annoying to restart the robot. They are required for Gazebo, however. -->
  <!-- Publish robot model -->
  <!-- <include file="$(find o2as_scene_description)/launch/o2as_base_scene_upload.launch" /> -->
  <!-- <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" /> -->

  <!-- Drivers for the robots -->
  <node name="a_bot_ur_driver"
	pkg ="ur_modern_driver" type="ur_driver"
	output="screen" ns="a_bot_controller">
    <param name="prefix"	   type="str"	 value="a_bot_" />
    <param name="robot_ip_address" type="str"	 value="$(arg robot_ip_a)" />
    <param name="reverse_port"	   type="int"	 value="50003" />
    <param name="min_payload"	   type="double" value="$(arg min_payload)" />
    <param name="max_payload"	   type="double" value="$(arg max_payload)" />
    <param name="max_velocity"	   type="double" value="$(arg max_velocity)" />
    <param name="servoj_time"	   type="double" value="$(arg servoj_time)" />
    <param name="base_frame"	   type="str"	 value="a_bot_base" />
    <param name="tool_frame"	   type="str"	 value="a_bot_tool0_controller" />
    <param name="require_activation"	 type="str" value="$(arg require_activation)" />
    <param name="shutdown_on_disconnect" type="bool" value="$(arg shutdown_on_disconnect)"/>
    <remap from="joint_states" to="/joint_states"/>
  </node>

  <node name="b_bot_ur_driver"
	pkg ="ur_modern_driver" type="ur_driver"
	output="screen" ns="b_bot_controller">
    <param name="prefix"	   type="str"	 value="b_bot_" />
    <param name="robot_ip_address" type="str"	 value="$(arg robot_ip_b)" />
    <param name="reverse_port"	   type="int"	 value="50002" />
    <param name="min_payload"	   type="double" value="$(arg min_payload)" />
    <param name="max_payload"	   type="double" value="$(arg max_payload)" />
    <param name="max_velocity"	   type="double" value="$(arg max_velocity)" />
    <param name="servoj_time"	   type="double" value="$(arg servoj_time)" />
    <param name="base_frame"	   type="str"	 value="b_bot_base" />
    <param name="tool_frame"	   type="str"	 value="b_bot_tool0_controller" />
    <param name="require_activation" type="str" value="$(arg require_activation)" />
    <param name="shutdown_on_disconnect" type="bool" value="$(arg shutdown_on_disconnect)"/>
    <remap from="joint_states" to="/joint_states"/>
  </node>

  <!-- Drivers for the grippers -->
  <!-- <include file="$(find robotiq_control)/launch/ur_cmodel_action_controller.launch"> -->
  <!--   <arg name="namespace"	value="a_bot_gripper" /> -->
  <!--   <arg name="gripper_prefix"	value="a_bot_" /> -->
  <!--   <arg name="topic"		value="/a_bot_controller/ur_driver/URScript" /> -->
  <!-- </include> -->

  <include file="$(find o2as_precision_gripper)/launch/precision_gripper.launch">
    <arg name="serial_port" default="/dev/ttyUSB0" />
  </include>

  <!-- Drivers for the force sensors -->
  <include file="$(find aist_ftsensor)/launch/aist_ftsensor.launch">
    <arg name="robot_name"	value="a_bot"/>
    <arg name="sensor_frame"	value="a_bot_ft300_wrench_link"/>
  </include>

  <include file="$(find aist_ftsensor)/launch/aist_ftsensor.launch">
    <arg name="robot_name"	value="b_bot"/>
    <arg name="sensor_frame"	value="b_bot_ft300_wrench_link"/>
  </include>

  <!-- Launch fastening tools -->
  <node name="suction_controller" pkg="o2as_fastening_tools"
	type="suction_controller.py" output="screen"
	unless="$(eval arg('sim'))">
    <param name="config_dir"	  type="string"	value="$(arg config_dir)"/>
    <param name="suction_control" type="string"	value="$(arg suction_control)"/>
    <param name="operation_mode"  type="string"	value="$(arg operation_mode)"/>
  </node>

</launch>
